stage1/init/init.go

```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+				symlinks[ln] = tgt
+			}
+		}
+	}
+
+	return symlinks
+}
+
+func createCgroups(root string, machineID string, appHashes []types.Hash) error {
+	cgroups, err := parseCgroups()
+	if err != nil {
+		return fmt.Errorf("error parsing /proc/cgroups: %v", err)
+	}
+
+	controllers := getControllers(cgroups)
+
+	// 1. Mount /sys read-only
+	sys := filepath.Join(root, "/sys")
+	if err := os.MkdirAll(sys, 0700); err != nil {
+		return err
+	}
+	if err := syscall.Mount("sysfs", sys, "sysfs", syscall.MS_RDONLY|syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV, ""); err != nil {
+		return fmt.Errorf("error mounting %q: %v", sys, err)
+	}
+
+	// 2. Mount /sys/fs/cgroup
+	cgroupTmpfs := filepath.Join(root, "/sys/fs/cgroup")
+	if err := os.MkdirAll(cgroupTmpfs, 0700); err != nil {
+		return err
+	}
+	if err := syscall.Mount("tmpfs", cgroupTmpfs, "tmpfs", syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV|syscall.MS_STRICTATIME, "mode=755"); err != nil {
+		return fmt.Errorf("error mounting %q: %v", cgroupTmpfs, err)
+	}
+
+	// 3. Mount controllers
+	for _, c := range controllers {
+		// 3a. Mount controller
+		cPath := filepath.Join(root, "/sys/fs/cgroup", c)
+		if err := os.MkdirAll(cPath, 0700); err != nil {
+			return err
+		}
+		if err := syscall.Mount("cgroup", cPath, "cgroup", syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV, c); err != nil {
+			return fmt.Errorf("error mounting %q: %v", cPath, err)
+		}
+
+		// 3b. Check if we're running from a unit to know which subcgroup
+		// directories to mount read-write
+		var subcgroup string
+		fromUnit, err := runningFromUnitFile()
+		if err != nil {
+			return fmt.Errorf("error determining if we're running from a unit file: %v", err)
+		}
+		if fromUnit {
+			slice, err := getSlice()
+			if err != nil {
+				return fmt.Errorf("error getting slice name: %v", err)
+			}
+			unit, err := getUnitFileName()
+			if err != nil {
+				return fmt.Errorf("error getting unit name: %v", err)
+			}
+			subcgroup = filepath.Join(cPath, slice, unit, "system.slice")
+		} else {
+			escapedmID := strings.Replace(machineID, "-", "\\x2d", -1)
+			machineDir := "machine-" + escapedmID + ".scope"
+			subcgroup = filepath.Join(cPath, "machine.slice", machineDir, "system.slice")
+		}
+
+		// 3c. Create and bind-mount app cgroup directories over themselves so
+		// they stay read-write
+		for _, a := range appHashes {
+			serviceName := ServiceUnitName(a)
+			appCgroup := filepath.Join(subcgroup, serviceName)
+
+			if err := os.MkdirAll(appCgroup, 0755); err != nil {
+				return err
+			}
+			if err := syscall.Mount(appCgroup, appCgroup, "", syscall.MS_BIND, ""); err != nil {
+				return fmt.Errorf("error bind mounting %q: %v", appCgroup, err)
+			}
+		}
+
+		// 3d. Re-mount controller read-only to prevent the container modifying host controllers
+		if err := syscall.Mount(cPath, cPath, "", syscall.MS_BIND|syscall.MS_REMOUNT|syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV|syscall.MS_RDONLY, ""); err != nil {
```

Comment:
general comment, can you split lines like this out (e.g. the flags) - otherwise it's way too easy to miss something in a review :-)